evaluation/plotting.py

```diff
@@ -42,87 +43,198 @@ def plot_bar_charts(
         axs = np.array([axs])
 
     for i, ax in enumerate(axs.flat):
-        x_data = list(data[i].keys())
-        y_data = list(data[i].values())
-        name = titles[i]
-        colors = plt.cm.tab20.colors[: len(x_data)]
+        ax = axs.flat[i]
+        category_labels = []
+        y_data = []
+        for key, categories in data.items():
+            category_labels.append(key)
+            all_x_data = set()
+            for category in categories:
+                all_x_data.update(category.keys())
+            all_x_data = sorted(list(all_x_data))
+            x_base = list(range(len(all_x_data)))
+            bar_width = 0.25
+            offsets = [-bar_width, 0, bar_width]
+            category_positions = [[x + offset for x in x_base] for offset in offsets]
+            bar_width = width / len(categories)
+            y_data_per_model = list(categories[i].values())
+            y_data.append(y_data_per_model)
+
+        colors = plt.cm.tab20.colors
+        for pos_list, height, color, label in zip(category_positions, y_data, colors, category_labels):
+            bars = ax.bar(pos_list, height, width=bar_width, color=color, label=label)
+            for bar in bars:
+                yval = bar.get_height()
+                ax.text(bar.get_x() + bar.get_width() / 2, yval, round(yval, 1), ha="center", va="bottom", fontsize=8)
 
-        ax.bar(x_data, y_data, width=width, label=name, color=colors)
-        ax.set_title(name, pad=20)
         if y_max_lim is not None and len(y_max_lim) > i and y_max_lim[i] is not None:
-            ax.set_ylim(0, max(y_max_lim[i], max(y_data)))
+            ax.set_ylim(0, y_max_lim[i] * 1.5)
```

Comment:
This looks problematic. Take a look at the generated evaluation results bar plots:
![evaluation_results](https://github.com/user-attachments/assets/b1875719-3009-4c0d-bc2e-eff42e3ef981)

What's up with the y-axis? Why do the percentages go above 1 (btw 0-1 is a possibility/probability, 0-100% is a percentage)? Why do the ratings (0-5) go all the way to 7? Why do we show more questions than what the evaluation actually ran on (that first plot is weird anyway, the percentage would be enough)?

Reply:
> This looks problematic. Take a look at the generated evaluation results bar plots: ![evaluation_results](https://private-user-images.githubusercontent.com/57714042/360467105-b1875719-3009-4c0d-bc2e-eff42e3ef981.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQzMzYxMDksIm5iZiI6MTcyNDMzNTgwOSwicGF0aCI6Ii81NzcxNDA0Mi8zNjA0NjcxMDUtYjE4NzU3MTktMzAwOS00YzBkLWJjMmUtZWZmNDJlM2VmOTgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIyVDE0MTAwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzZjMwMmQ2ZmViOTkxYjZmMGY2ZmJmZGU1NjczOWY0Yzc5YzIzOWFiODFkMzZjNjE1YjAwNDEzODZhZjZhZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.w7z2F3WPk5_izLbB2Q1XZzdjCy6yeI4L9ZErihhRZqQ)
> 
> What's up with the y-axis? Why do the percentages go above 1 (btw 0-1 is a possibility/probability, 0-100% is a percentage)? Why do the ratings (0-5) go all the way to 7? Why do we show more questions than what the evaluation actually ran on (that first plot is weird anyway, the percentage would be enough)?

I did this to avoid the legend overlapping with the bars. But yeah the y-axis values don't make sense this way.

Reply:
Got it! Let's find a way to do that without changing the y-axis! Can we perhaps have a single legend for all 3 diagrams outside of them?